[HUDI-9013] Add backwards compatible MDT writer support and reader support with tbl v6

lokeshj1703 opened the pull request:
### Change Logs

With 1.0 release, we did not add support for b/w compatible support for MDT. 

We wanted to add backwards compatible MDT writer support with 1.x binary to read and write MDT w/ tbl version 6. With this, 1.0 binary should be able to read tables w/ MDT enabled from 0.x and 1.x with tbl version 6. Also, we should be able to write to MDT using tbl v 6 with 1.0 binary. 

### Impact

Adds support for backward compatible metadata writer with table version 6 in 1.0.

### Risk level (write none, low medium or high below)

low

### Documentation Update

NA

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

lokeshj1703 commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/metadata/FlinkHoodieBackedTableMetadataWriter.java:
```diff
@@ -133,7 +133,7 @@ protected void commitInternal(String instantTime, Map<String, HoodieData<HoodieR
     }
 
-    compactIfNecessary(writeClient);
+    metadataWriteHandler.compactIfNecessary(writeClient, "");
```
No changes are made in Flink MW

danny0405 replied:
We are assuming Flink table version is always above 6?

lokeshj1703 replied:
Addressed

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/BaseTableServicePlanActionExecutor.java:
```diff
@@ -124,7 +124,9 @@ public Pair<Option<HoodieInstant>, Set<String>> getIncrementalPartitions(TableSe
             String completionTime = instant.getCompletionTime();
-            if (completionTime.compareTo(leftBoundary) >= 0 && completionTime.compareTo(rightBoundary) < 0) {
+            if (completionTime.compareTo(leftBoundary) >= 0
+                && ((instant.requestedTime().length() < completionTime.length() && instant.requestedTime().compareTo(rightBoundary) < 0)
```
This line is confusing, can we at least add a comment, or maybe abstract it as a separate method.

nsivabalan replied:
yes, and lets add java docs.

lokeshj1703 replied:
Not part of this PR. Removed it.

danny0405 commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/SparkBroadcastManager.java:
```diff
@@ -116,10 +119,8 @@ public Option<HoodieReaderContext> retrieveFileGroupReaderContext(StoragePath ba
-          sparkParquetReader,
-          JavaConverters.asScalaBufferConverter(filters).asScala().toSeq(),
-          JavaConverters.asScalaBufferConverter(filters).asScala().toSeq()));
+      return Option.of(new SparkFileFormatInternalRowReaderContext(sparkParquetReader, JavaConverters.asScalaBufferConverter(filters).asScala().toSeq(),
```
Maybe we follow the original indentation.

lokeshj1703 replied:
Not part of this PR. Removed it.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieBaseFileGroupRecordBuffer.java:
```diff
@@ -424,8 +424,10 @@ protected Option<T> merge(Option<T> older, Map<String, Object> olderInfoMap,
-          if (!isDeleteRecordWithNaturalOrder(older, oldOrderingValue)
-              && oldOrderingValue.compareTo(newOrderingValue) > 0) {
+          boolean choosePrev = !oldOrderingValue.equals(0)
+              && ReflectionUtils.isSameClass(oldOrderingValue, newOrderingValue)
```
Is this a bug fix for master?

lokeshj1703 replied:
Not part of this PR. Removed it.

danny0405 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/HoodieCombineRealtimeRecordReader.java:
```diff
@@ -54,8 +54,8 @@ public class HoodieCombineRealtimeRecordReader implements RecordReader<NullWrita
 
   public HoodieCombineRealtimeRecordReader(JobConf jobConf, CombineFileSplit split,
       List<RecordReader> readers) {
+    useFileGroupReader = shouldUseFilegroupReader(jobConf, split);
```
why this change

lokeshj1703 replied:
Not part of this PR. Removed it.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -155,6 +160,8 @@ public abstract class HoodieBackedTableMetadataWriter<I> implements HoodieTableM
   protected StorageConfiguration<?> storageConf;
   protected final transient HoodieEngineContext engineContext;
   protected final List<MetadataPartitionType> enabledPartitionTypes;
+  protected final MetadataWriteHandler metadataWriteHandler;
```
can you add java docs on the purpose of write handler

danny0405 replied:
I feel like maybe we just sub-class the whole `HoodieBackedTableMetadataWriter` instead of introducing `MetadataWriteHandler` on existing code, this makes the code clean and more easier to remove the legacys in the future.

nsivabalan replied:
yeah, we also wanted to do that. 
but due to our hierarchy, we could not come up w/ an elegant way to do that. 

bcoz, we have HoodieBackedTableMetadatWriter. and Flink, Java and SparkBackedTableMetadatWriter extends from this. 

So, not easy to make entire class HoodieBackedTableMetadatWriter as V1 and V2. So, we went w/ introducing a new write handler internally within HoodieBackedTableMetadatWriter. 

Let us know if you have better ideas.

danny0405 replied:
We can add a base extended class from `HoodieBackedTableMetadatWriter`, then share it with different engines.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1726,4 +1659,302 @@ public boolean isInitialized() {
+     * 2. In multi-writer scenario, a parallel operation with a greater instantTime may have completed creating a
+     * deltacommit.
+     */
+    protected void compactIfNecessary(BaseHoodieWriteClient writeClient, String latestDeltacommitTime) {
```
this java docs need to go after L 1775.

lokeshj1703 replied:
Java docs for the function `compactIfNecessary` are same in 0.15.0 and 1.0

nsivabalan commented:
@danny0405 : he is reverting the changes he did earlier to disable file group reader for table version 6. 
This has a separate patch https://github.com/apache/hudi/pull/12935  and Lokesh is stacking this patch on top of that.

lokeshj1703 commented:
Added v6 test for TestRecordLevelIndex, TestColumnStatsIndex and TestHoodieBackedTableMetadata. Added older instant time generation logic to keep it consistent with 0.15.0 for table version 6.

lokeshj1703 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1726,4 +1659,302 @@ public boolean isInitialized() {
+     * 2. In multi-writer scenario, a parallel operation with a greater instantTime may have completed creating a
+     * deltacommit.
+     */
+    protected void compactIfNecessaryVersionSix(BaseHoodieWriteClient writeClient, String latestDeltacommitTime) {
```
Java docs for the function `compactIfNecessary` are same in 0.15.0 and 1.0

lokeshj1703 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -344,9 +355,13 @@ private boolean isBootstrapNeeded(Option<HoodieInstant> latestMetadataInstant) {
-  private void initializeFromFilesystem(String initializationTime, List<MetadataPartitionType> partitionsToInit) throws IOException {
+  private boolean initializeFromFilesystem(String initializationTime, List<MetadataPartitionType> partitionsToInit, Option<String> inflightInstantTimestamp) throws IOException {
     Set<String> pendingDataInstants = getPendingDataInstants(dataMetaClient);
+    if (!metadataWriteHandler.shouldInitializeFromFilesystem(pendingDataInstants, inflightInstantTimestamp)) {
```
I am now returning false here. The behaviour is inline with table version 6 now.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1734,4 +1654,387 @@ public boolean isInitialized() {
 
   protected abstract BaseHoodieWriteClient<?, I, ?, ?> initializeWriteClient();
+
+  protected class MetadataWriteHandler implements Serializable {
```
java docs.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1734,4 +1654,387 @@ public boolean isInitialized() {
+
+  protected class MetadataWriteHandler implements Serializable {
+    private static final int PARTITION_INITIALIZATION_TIME_SUFFIX = 10;
+    HoodieTableVersion tableVersion;
```
private final

lokeshj1703 replied:
Addressed

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1734,4 +1654,387 @@ public boolean isInitialized() {
+      }
+    }
+
+    public String getTimelineHistoryPath() {
```
can you revisit all access specifiers

nsivabalan commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/functional/TestHoodieBackedTableMetadata.java:
```diff
@@ -101,10 +102,15 @@ public class TestHoodieBackedTableMetadata extends TestHoodieMetadataBase {
+    HoodieWriteConfig config = getWriteConfigBuilder(HoodieFailedWritesCleaningPolicy.EAGER, true, true, false, true, false)
+        .build();
+    config.setValue(HoodieTableConfig.VERSION, String.valueOf(tableVersion));
+    config.setValue(HoodieWriteConfig.WRITE_TABLE_VERSION, String.valueOf(tableVersion));
```
we only need to set the `write table version`. can you avoid setting the `table version` directly.

nsivabalan replied:
at the end of every test, can we also add validation of table version in table config in meta client.

lokeshj1703 replied:
Addressed. There were few modifications required for it in HoodieTestUtils.

nsivabalan commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestRecordLevelIndex.scala:
```diff
@@ -418,48 +419,6 @@ class TestRecordLevelIndex extends RecordLevelIndexTestBase {
-  @CsvSource(value = Array(
-    "COPY_ON_WRITE,6", "COPY_ON_WRITE,8", "MERGE_ON_READ,6", "MERGE_ON_READ,8"
-  ))
-  def testRLIWithDTClustering(tableType: String, tableVersion: Int): Unit = {
```
why deleted this test?

lokeshj1703 replied:
This was an unintended change. Added it back.

danny0405 commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestColumnStatsIndex.scala:
```diff
@@ -847,7 +856,7 @@ class TestColumnStatsIndex extends ColumnStatIndexTestBase {
       .save(basePath)
 
-    metaClient = HoodieTableMetaClient.reload(metaClient)
+    metaClient = HoodieTableMetaClient.builder().setBasePath(basePath).setConf(storageConf).build()
```
why does the meta client reload not work now?

danny0405 commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestColumnStatsIndex.scala:
```diff
@@ -847,7 +856,7 @@ class TestColumnStatsIndex extends ColumnStatIndexTestBase {
       .save(basePath)
 
-    metaClient = HoodieTableMetaClient.reload(metaClient)
+    metaClient = HoodieTableMetaClient.builder().setBasePath(basePath).setConf(storageConf).build()
```
why does the meta client reload not work now?

lokeshj1703 replied:
The new table configs are not reflected with reload. Reload takes some arguments from old meta client (created in harness utility class I think). Those properties do not work here.

danny0405 replied:
Maybe we just copy the tests instead of modification on existing one, the additional version param makes the existing tests much more complicated and hard to maintain.

danny0405 commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestColumnStatsIndexTableVersionSix.scala:
```diff
@@ -0,0 +1,31 @@
+class TestColumnStatsIndexTableVersionSix extends TestColumnStatsIndex {
+
+  override val overrideOpts: Map[String, String] = Map(
+    HoodieTableConfig.VERSION.key() -> "6",
```
we can just copy the test methods from `TestColumnStatsIndex` that require the `overrideOpts` to this class though and the method with version param added so that there is no need to to change the `TestColumnStatsIndex `.

nsivabalan replied:
we should avoid copying the entire test class.

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriterTableVersionSix.java:
```diff
@@ -0,0 +1,337 @@
+ * HoodieBackedTableMetadataWriter for tables with version 6. The class derives most of the functionality from HoodieBackedTableMetadataWriter
+ * and overrides some behaviour to make it compatible for version 6.
+ */
+public abstract class HoodieBackedTableMetadataWriterTableVersionSix<I> extends HoodieBackedTableMetadataWriter<I> {
```
Can we merge it with its Spark extendee, it looks like Flink and Java do not need it.

lokeshj1703 replied:
If I merge, it is causing issues with some APIs like compactIfNecessary

nsivabalan commented:
CI is failing due to known flaky test ITTestHoodieDataSource. testIncrementalReadArchivedCommits

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -155,8 +155,9 @@ public abstract class HoodieBackedTableMetadataWriter<I> implements HoodieTableM
+
   // Is the MDT bootstrapped and ready to be read from
-  private boolean initialized = false;
+  boolean initialized = false;
```
why not protected or private?

lokeshj1703 replied:
Since the classes are in the same package, the default package private modifier is sufficient here.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -193,6 +194,10 @@ protected HoodieBackedTableMetadataWriter(StorageConfiguration<?> storageConf,
     ValidationUtils.checkArgument(!initialized || this.metadata != null, "MDT Reader should have been opened post initialization");
   }
 
+  List<MetadataPartitionType> getEnabledPartitions(TypedProperties writeConfigProps, HoodieTableMetaClient metaClient) {
```
again, lets try to ensure access specifiers are tight

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1257,6 +1278,10 @@ public void update(HoodieRestoreMetadata restoreMetadata, String instantTime) {
     }
   }
 
+  String createRestoreInstantTime() {
```
protected

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1552,7 +1577,7 @@ static HoodieActiveTimeline runPendingTableServicesOperationsAndRefreshTimeline(
    * deltacommit.
    */
-  protected void compactIfNecessary(BaseHoodieWriteClient writeClient) {
+  void compactIfNecessary(BaseHoodieWriteClient<?,I,?,?> writeClient, String latestDeltacommitTime) {
```
why removed protected?

nsivabalan commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/metadata/FlinkHoodieBackedTableMetadataWriter.java:
```diff
@@ -133,7 +133,7 @@ protected void commitInternal(String instantTime, Map<String, HoodieData<HoodieR
     }
 
-    compactIfNecessary(writeClient);
+    compactIfNecessary(writeClient, "");
```
can we make 2nd arg as Option<String> 
for table version 6, we can validate that this arg is present

lokeshj1703 replied:
Addressed

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedTableMetadataWriterTableVersionSix.java:
```diff
@@ -0,0 +1,164 @@
+
+  @Override
+  protected void initRegistry() {
+    if (metadataWriteConfig.isMetricsOn()) {
```
why overriding this? is this diff in v 6 and v8?

lokeshj1703 replied:
yeah..it is a copy of `SparkHoodieBackedTableMetadataWriter`

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedTableMetadataWriterTableVersionSix.java:
```diff
@@ -0,0 +1,164 @@
+  }
+
+  @Override
+  public void deletePartitions(String instantTime, List<MetadataPartitionType> partitions) {
```
oh its a complete copy paste of SparkHoodieBackedTableMetadataWriter.

nsivabalan commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieTestUtils.java:
```diff
@@ -233,6 +233,10 @@ public static HoodieTableMetaClient.TableBuilder getMetaClientBuilder(HoodieTabl
       builder.setKeyGeneratorType(properties.getProperty(HoodieTableConfig.KEY_GENERATOR_TYPE.key()));
     }
 
+    if (properties.containsKey("hoodie.write.table.version")) {
```
lets try to use the variables already defined for this. also in L 237

lokeshj1703 replied:
The config variables are not available in this module

nsivabalan commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestRecordLevelIndexTableVersionSix.scala:
```diff
@@ -0,0 +1,29 @@
+import org.apache.hudi.config.HoodieWriteConfig
+
+class TestRecordLevelIndexTableVersionSix extends TestRecordLevelIndex {
+  override def commonOpts: Map[String, String] = super.commonOpts ++ Map(
```
why can't we do something similar for col stats test?

lokeshj1703 replied:
Addressed. Parameterised the tests.
